### ISSUE TITLE
IPNS notifications

### ISSUE THREAD

**Issue by ajres** [ipfs/ipfs#73]
Is there a mechanism for a Node to receive events when an IPNS of interest is updated, other than polling?

Andy

**Comment by harlantwood** [comment]
+1 very interested to know about this if it exists, or upvote the feature if not.  @john-light I believe this would form a foundation for realtime chat, as well as a first of other applications, as we were just discussing...

**Comment by jbenet** [comment]
- yeah we really want this.
- we _will be_ implementing a pubsub mechanism that uses ipns + the routing system
- pubsub gives you push notifications
- not clear how it will work exactly yet, but several solutions have been sketched out. 

just one example, but by no means the best:

- store a "pubsub group" record under an ipns name.
- people subscribe to it by storing "pubsub subscribe" routing records at the group's key. (DSHT so can store many records per key)
- publishers would periodically check for subscriptions, keep a local list.
- (optionally) publishers gossip (like PEX) subscriber records to ensure they have a complete view
- publishers send notifications directly to subscribers.

other ways of doing this can perform much better (rely less on high perf of routing) if we have the subscribers or the publishers do a bit more work (maintaining the table or even relaying messages)

**Comment by gjvnq** [comment]
I think that sending notifications directly to the nodes is a bad idea because the node may be powered off when the message is sent.

One possible way around this is for nodes to publish requests like they would do with files but with backlinks to the channel IPNS address.

When the channel owner wants to send messages, they look for all requests that backlinks to the channel IPNS address and publish new messages (in a simlar way blocks are published) linking to the recipients.

A while later, when a node wants to checks its messages, it search for all messages that link to it.

**Comment by fps** [comment]
Hi, while it is not true pub/sub in the sense that a node gets a message about a change, but rather it involves polling, morphis has dpush, which is the technology underlying morphis' dmail. It basically consists of the publication of a target key prefix which the node will regularly scan using the prefix search inherent in kademlia.

it also depends on morphis' notion of updateable keys (that is documents can be updated while retaining their key). so i don't know how applicable it is to ipfs, but maybe there is enough overlap. here's some links for further details:

https://morph.is/
https://morph.is/v0.8/faq.html

and especially:

https://morph.is/v0.8/dpush-whitepaper.odt

**Comment by davidar** [comment]
> it also depends on morphis' notion of updateable keys

IP**N**S would be the equivalent here.

**Comment by DaneBettis** [comment]
I am also interested in this feature: a pub/sub system, even if not really real-time, would enable a robust VFX asset usage pipeline.

I'd be glad to further discuss options and possibly provide some real-world testing of ipfs in the VFX content development production pipeline.

**Comment by doublethefish** [comment]
I was thinking the same thing.

**Comment by 1l0** [comment]
:+1:

**Comment by jbenet** [comment]
Indeed. We're a bit overloaded with things atm. if you want to speed it up, coordinate here https://github.com/ipfs/notes/issues/64
